src/libraries/ProtocolFeeLibrary.sol

```diff
@@ -1,42 +1,42 @@
 // SPDX-License-Identifier: GPL-2.0-or-later
 pragma solidity ^0.8.20;
 
-import "./UnsafeMath.sol";
-
 library ProtocolFeeLibrary {
     // Max protocol fee is 0.1% (1000 pips)
     uint16 public constant MAX_PROTOCOL_FEE = 1000;
 
+    // Thresholds used for optimized bounds checks on protocol fees
+    uint24 internal constant FEE_0_THRESHOLD = 1001;
+    uint24 internal constant FEE_1_THRESHOLD = 1001 << 12;
+
     // the protocol fee is represented in hundredths of a bip
     uint256 internal constant PIPS_DENOMINATOR = 1_000_000;
 
     function getZeroForOneFee(uint24 self) internal pure returns (uint16) {
-        return uint16(self & (4096 - 1));
+        return uint16(self & 0xfff);
     }
 
     function getOneForZeroFee(uint24 self) internal pure returns (uint16) {
         return uint16(self >> 12);
     }
 
-    function isValidProtocolFee(uint24 self) internal pure returns (bool) {
-        if (self != 0) {
-            uint16 fee0 = getZeroForOneFee(self);
-            uint16 fee1 = getOneForZeroFee(self);
-            // The fee is represented in pips and it cannot be greater than the MAX_PROTOCOL_FEE.
-            if ((fee0 > MAX_PROTOCOL_FEE) || (fee1 > MAX_PROTOCOL_FEE)) {
-                return false;
-            }
+    function isValidProtocolFee(uint24 self) internal pure returns (bool success) {
+        // Equivalent to: self == 0 ? true : (getZeroForOneFee(self) <= MAX_PROTOCOL_FEE && getOneForZeroFee(self) <= MAX_PROTOCOL_FEE)
+        assembly {
+            let isZeroForOneFeeOk := lt(and(self, 0xfff), FEE_0_THRESHOLD)
+            let isOneForZeroFeeOk := lt(self, FEE_1_THRESHOLD)
+            success := or(iszero(self), and(isZeroForOneFeeOk, isOneForZeroFeeOk))
```

Comment:
couldnt this line just be `and(isZeroForOneFeeOk, isOneForZeroFeeOk)` ?
if `iszero(self)`, both `isZeroForOneFeeOk` and `isOneForZeroFeeOk` will be true.
I think the `self != 0` was just a gas optimisation previously so that if the fee is 0 you dont have to do anything at all

Reply:
Good catch!